toolchain/check/import_ref.cpp

```diff
@@ -224,6 +333,145 @@ class ImportRefResolver {
+
+    // Build the `Self` type.
+    auto type_id = context_.GetClassType(class_decl.class_id);
+    context_.classes().Get(class_decl.class_id).self_type_id = type_id;
+
+    // Set a constant corresponding to the incomplete class.
+    // TODO: This uses the self type's constant, but the underlying instruction
+    // is NotConstant. Need to discuss whether we want to add a type_id to
+    // ClassDecl/InterfaceDecl and change behavior, which would also help
+    // consistency with other types. See also GetExprValueForLookupResult in
+    // handle_name.cpp.
+    auto const_id = context_.types().GetConstantId(type_id);
+    import_ir_constant_values_.Set(inst_id, const_id);
+    return const_id;
+  }
+
+  // Fills out the class definition for an incomplete class.
+  auto AddClassDefinition(const SemIR::Class& import_class,
+                          SemIR::ConstantId class_const_id,
+                          SemIR::ConstantId object_repr_const_id,
+                          SemIR::ConstantId base_const_id) -> void {
+    auto& new_class = context_.classes().Get(
+        context_.insts()
+            .GetAs<SemIR::ClassType>(class_const_id.inst_id())
+            .class_id);
+
+    new_class.object_repr_id =
+        context_.GetTypeIdForTypeConstant(object_repr_const_id);
+
+    new_class.scope_id =
+        context_.name_scopes().Add(new_class.decl_id, SemIR::NameId::Invalid,
+                                   new_class.enclosing_scope_id);
+    auto& new_scope = context_.name_scopes().Get(new_class.scope_id);
+    const auto& old_scope = import_ir_.name_scopes().Get(import_class.scope_id);
+    // Push a block so that we can add scoped instructions to it, primarily for
+    // textual IR formatting.
+    context_.inst_block_stack().Push();
+    for (auto [entry_name_id, entry_inst_id] : old_scope.names) {
+      CARBON_CHECK(
+          new_scope.names
+              .insert({GetLocalNameId(entry_name_id),
+                       context_.AddPlaceholderInst(SemIR::ImportRefUnused{
+                           import_ir_id_, entry_inst_id})})
+              .second);
+    }
+    new_class.body_block_id = context_.inst_block_stack().Pop();
+
+    if (import_class.base_id.is_valid()) {
+      new_class.base_id = base_const_id.inst_id();
+      // Add the base scope to extended scopes.
+      auto base_inst_id = context_.types().GetInstId(
+          context_.insts()
+              .GetAs<SemIR::BaseDecl>(new_class.base_id)
+              .base_type_id);
+      const auto& base_class = context_.classes().Get(
+          context_.insts().GetAs<SemIR::ClassType>(base_inst_id).class_id);
+      new_scope.extended_scopes.push_back(base_class.scope_id);
+    }
+    CARBON_CHECK(new_scope.extended_scopes.size() ==
+                 old_scope.extended_scopes.size());
+  }
+
+  auto TryResolveTypedInst(SemIR::ClassDecl inst, SemIR::InstId inst_id,
+                           bool incomplete_type) -> SemIR::ConstantId {
```

Comment:
And here, since we're also manipulating the member of `Work`, I wonder if there is a different name for the function variable to indicate the separation?

`already_resolved_incomplete_type` is awfully long.

Alternatively, could consider a different naming structure: `pre_incomplete_type` to indicate the first phase?

Or to keep `false`-by-default, maybe `type_second_pass`? And then these comments can remind the reader that the second pass doesn't need to create the incomplete type but does need to load. WDYT? Happy to chat about naming options if none of these work.

Reply:
As above, `made_incomplete_type`

`type_second_pass` isn't quite the semantic I was aiming for because we still might need to check the constant to see if the instruction was previously filled in for some other edge cases.